File: app/lib/models/bracket.dart
class BracketBook {
  final String id;
  final String title;
  final String author;
  final String coverColor;
  final String? coverUrl;
  final double? rating;
  final int? month;

  BracketBook({
    required this.id,
    required this.title,
    required this.author,
    required this.coverColor,
    this.coverUrl,
    this.rating,
    this.month,
  });
}

class BracketFavorite {
  final int month;
  final BracketBook? book;

  BracketFavorite({required this.month, this.book});
}

class BracketMatch {
  final String id;
  final String round;
  final BracketBook? bookA;
  final BracketBook? bookB;
  final String? winnerId;

  BracketMatch({required this.id, required this.round, this.bookA, this.bookB, this.winnerId});

  bool get isReady => bookA != null && bookB != null;
}

class BracketData {
  final int year;
  final int monthsSet;
  final List<BracketFavorite> favorites;
  final List<BracketMatch>? matches;
  final BracketBook? champion;

  BracketData({
    required this.year,
    required this.monthsSet,
    required this.favorites,
    this.matches,
    this.champion,
  });

  bool get isComplete => monthsSet == 12;

  BracketMatch? matchById(String id) {
    if (matches == null) return null;
    for (final m in matches!) {
      if (m.id == id) return m;
    }
    return null;
  }

  List<BracketMatch> byRound(String round) => (matches ?? []).where((m) => m.round == round).toList();
}

File: backend/app/schemas/bracket.py
import uuid

from pydantic import BaseModel, Field


class MonthlyFavoriteSet(BaseModel):
    book_id: uuid.UUID


class BracketWinnerSet(BaseModel):
    book_id: uuid.UUID


class BracketBookOut(BaseModel):
    id: uuid.UUID
    title: str
    author: str
    cover_color: str
    cover_url: str | None
    rating: float | None
    month: int | None = None


class BracketFavoriteOut(BaseModel):
    month: int
    book: BracketBookOut | None


class BracketMatchOut(BaseModel):
    id: str
    round: str
    book_a: BracketBookOut | None
    book_b: BracketBookOut | None
    winner_id: uuid.UUID | None


class BracketOut(BaseModel):
    year: int
    months_set: int = Field(description="How many of the 12 months have a favourite picked")
    favorites: list[BracketFavoriteOut]
    matches: list[BracketMatchOut] | None
    champion: BracketBookOut | None
